als-common/shared/src/main/scala/org/mulesoft/als/common/YPartBranch.scala

```diff
@@ -1,18 +1,113 @@
 package org.mulesoft.als.common
 
-import amf.core.client.common.position.{Position => AmfPosition}
 import amf.core.client.scala.model.document.BaseUnit
 import amf.core.internal.parser.YNodeLikeOps
-import org.mulesoft.als.common.YamlWrapper._
+import org.mulesoft.als.common.ASTElementWrapper._
+import org.mulesoft.als.common.YPartASTWrapper._
 import org.mulesoft.als.common.dtoTypes.Position
 import org.mulesoft.amfintegration.AmfImplicits.{AmfAnnotationsImp, BaseUnitImp}
+import org.mulesoft.antlrast.ast.{ASTNode, Node, Terminal}
+import org.mulesoft.common.client.lexical.{ASTElement, Position => AmfPosition}
 import org.yaml.lexer.YamlCharRules
 import org.yaml.model
 import org.yaml.model._
 
 import scala.annotation.tailrec
 
-case class YPartBranch(node: YPart, position: AmfPosition, stack: Seq[YPart], isJson: Boolean, isInFlow: Boolean) {
+trait ASTPartBranch {
+  type T <: ASTElement
+  val stack: Seq[T]
+  val node: T
+  lazy val isMultiline = false
+  val isEmptyNode: Boolean
+  val stringValue: String
+  def keys: Seq[String]
+  val isKey: Boolean
+  val isValue: Boolean
+  val isAtRoot: Boolean
+  val isArray: Boolean
+  lazy val isInArray: Boolean = false
+  val isJson: Boolean
```

Comment:
is this OK at this level? or should it be at the lower abstraction level with T = YPart